README
ICanHasCheezburger.com addon for viewing LOL cats and more in XBMC
created by Brian Millham <brian@millham.net>

Licensed under GPL (see LICENSE.txt for details)


changelog.txt
ICanHasCheezburger.com Add-On For XBMC Changelog

Version 0.1: Initial version ported from the original plugin.

-- 08/20/2010: Created Branch to start developing the ability to see current LOLs

Version 0.2: (8/21/2010) Initial support of current LOLs (cats only)



addon.py
##############################################################################
# ICanHasCheezBurger = XBMC Addon
#
# based on Comics.com - XBMC picture plugin by Dan Dar3
#
# Coding by Brian Millham <brian@millham.net>
#
# Credits:
#   * Team XBMC @ XBMC.org                                                [http://xbmc.org/]
#   * Leonard Richardson <leonardr@segfault.org>  - BeautifulSoup 3.0.7a  [http://www.crummy.com/software/BeautifulSoup/]
#   * Eric Lawrence      <e_lawrence@hotmail.com> - Fiddler Web Debugger  [http://www.fiddler2.com]

import os
import sys

LIB_DIR = xbmc.translatePath( os.path.join( os.getcwd(), 'resources', 'lib' ) )
sys.path.append (LIB_DIR)

import xbmcaddon

__settings__ = xbmcaddon.Addon(id='plugin.image.icanhascheezburger.com')
__language__ = __settings__.getLocalizedString

if ( "action=list" in sys.argv[ 2 ] ):
    import random_cheez_list as plugin
    plugin.Main()
elif ("action=30405" in sys.argv[2]):
    import current_cheez as plugin
    plugin.Main()
elif ("action=30406" in sys.argv[2]):
    import random_cheez_selection as plugin
    plugin.Main()
else:
    xbmc.log( "[PLUGIN] %s v%s (%s)" % ( __plugin__, __version__, __date__ ), xbmc.LOGNOTICE )
    import cheezburger_type_selection as plugin
    plugin.Main()


resources/lib/xbmcplugin_main.py
# ICanHasCheezBurger Random Lol XBMC Plugin
# Based on Dan Dare's Comics.com plugin
# 
# Modified from the original Comics.com plugin
#  by Brian Millham <brian@millham.net>
#
# Imports
#
import sys
import xbmcgui
import xbmcplugin
import urllib

import xbmcaddon

__settings__ = xbmcaddon.Addon(id='plugin.pictures.icanhascheezburger.com')

__language__ = __settings__.getLocalizedString

#
# Main class
#
class Main:
    #
    # Init
    #
    def __init__(self):
        #
        # Get Lols
        #
        self.getLols( )
        
    #
    # Get Lols
    #
    def getLols(self):
        url_base = "http://api.cheezburger.com/xml/category/%s/lol/random/"
        #lol_list = sorted(['Cats', 'Dogs', 'Objects',
        #                   'Other Animals', 'News', 'Celebs',
        #                   'Fail', 'Engrish', 'Comix'] )
        lol_map = {30501: 'Cats',
                   30502: 'Dogs',
                   30503: 'Objects',
                   30504: 'Other Animals',
                   30505: 'News',
                   30506: 'Celebs',
                   30507: 'Fail',
                   30508: 'Engrish',
                   30509: 'Comix'}
        lol_list = []
        for s in range(30501, 30510):
            lol_list.append((__language__(s), lol_map[s]))
        for lol in lol_list:
            lol_name = __language__(30404) % lol[0]
            lol_url = url_base % lol[1]
            plugin_url = "%s?action=list&lol_name=%s&lol_url=%s" % (
             sys.argv[0],
             urllib.quote_plus( lol_name ),
             urllib.quote_plus( lol_url ) )
            #
            #  Add list item...
            # 
            listitem = xbmcgui.ListItem (lol_name,
             iconImage = "DefaultPicture.png",
             thumbnailImage = "DefaultPicture.png")
            listitem.setInfo( "pictures", { "title" : "lol_name" } )
            ok = xbmcplugin.addDirectoryItem( handle = int(sys.argv[1]),
             url = plugin_url,
             listitem = listitem,
             isFolder = True)

        #
        # End of list...
        #

        # Sort by label...
        xbmcplugin.addSortMethod( handle=int( sys.argv[ 1 ] ), sortMethod=xbmcplugin.SORT_METHOD_LABEL )
            
        # End of list...
        xbmcplugin.endOfDirectory( handle=int( sys.argv[ 1 ] ), succeeded=True )        


resources/lib/random_cheez_list.py
#
# Imports
#
import os
import sys
import xbmc
import xbmcgui
import xbmcplugin
import urllib
from BeautifulSoup    import BeautifulStoneSoup
from xbmcplugin_utils import HTTPCommunicator

import xbmcaddon

__settings__ = xbmcaddon.Addon(id='plugin.pictures.icanhascheezburger.com')
__language__ = __settings__.getLocalizedString

#
# Main class
#
class Main:
    #
    # Init
    #
    def __init__( self ) :
        # Constants
        self.DEBUG            = False
        self.IMAGES_PATH      = xbmc.translatePath( os.path.join( os.getcwd(), 'resources', 'images' ) )
        
        # Parse parameters...
        params = dict(part.split('=') for part in sys.argv[ 2 ][ 1: ].split('&'))
        self.lol_name       = urllib.unquote_plus( params[ "lol_name" ] )
        self.lol_url        = urllib.unquote_plus( params[ "lol_url" ] )
        self.current_page     =                int ( params.get( "page", "1" ) )
        self.entries_per_page = 5
        # Parameters
        if __settings__.getSetting ("entries_per_page") == "0" :
            self.entries_per_page = 5
        elif __settings__.getSetting ("entries_per_page") == "1" :
            self.entries_per_page = 10
        elif __settings__.getSetting ("entries_per_page") == "2" :
            self.entries_per_page = 20

        #
        # Get picture list...
        #
        self.getPictures()
    
    #
    # Get pictures...
    #
    def getPictures( self ) :        
        #
        # Get HTML page...
        #
        httpCommunicator = HTTPCommunicator()
        htmlSource = httpCommunicator.get( self.lol_url + str(self.entries_per_page))

        #
        # Parse HTML page...
        #
        pictures = BeautifulStoneSoup(htmlSource).findAll('picture')
        for picture in pictures:
            #
            # Title (date)
            #
            title       = picture.title.string
           
            #
            # Thumbnail & full image...
            #

            full_image_url = picture.lolimageurl.string
            thumbnail_url  = picture.thumbnailimageurl.string
            
                                    
            # Add directory entry...
            listitem = xbmcgui.ListItem( title, iconImage="DefaultPicture.png", thumbnailImage = thumbnail_url )
            xbmcplugin.addDirectoryItem( handle=int(sys.argv[ 1 ]), url = full_image_url, listitem=listitem, isFolder=False)

        #
        # Next page entry...
        #
        next_button_text = __language__(30403) % (self.entries_per_page, self.lol_name)
        listitem      = xbmcgui.ListItem (next_button_text, iconImage = "DefaultFolder.png", thumbnailImage = os.path.join(self.IMAGES_PATH, 'next-page.png'))
        next_page_url = "%s?action=list&lol_name=%s&lol_url=%s" % ( sys.argv[0], urllib.quote_plus( self.lol_name), urllib.quote_plus( self.lol_url ) )
        #next_page_url = self.lol_url
        xbmcplugin.addDirectoryItem( handle = int(sys.argv[1]), url = next_page_url, listitem = listitem, isFolder = True)
        
        #    
        # Disable sorting...
        #
        xbmcplugin.addSortMethod( handle=int( sys.argv[ 1 ] ), sortMethod=xbmcplugin.SORT_METHOD_NONE )
        
        #
        # Label (top-right)...
        #
        trlabel = "%s (%s)" % ( self.lol_name, (__language__(30402) % self.current_page))
        xbmcplugin.setPluginCategory( handle=int( sys.argv[ 1 ] ), category=( trlabel ) )
        
        #
        # End of directory...
        #
        xbmcplugin.endOfDirectory( handle=int( sys.argv[ 1 ] ), succeeded=True )


resources/lib/current_cheez_list.py
__author__="brian"
__date__ ="$Aug 20, 2010 11:11:39 PM$"

# Show current LOLs. For now, only Cats are supported. Others will follow.

import os
import sys
import xbmc
import xbmcgui
import xbmcplugin
import feedparser
import urllib
from BeautifulSoup    import BeautifulStoneSoup
import xbmcaddon

__settings__ = xbmcaddon.Addon(id='plugin.pictures.icanhascheezburger.com')
__language__ = __settings__.getLocalizedString

class Main:
    #
    # Init
    #
    def __init__(self, url) :
        # Constants
        self.DEBUG            = False
        self.IMAGES_PATH      = xbmc.translatePath( os.path.join( os.getcwd(), 'resources', 'images' ) )
        #params = dict(part.split('=') for part in sys.argv[ 2 ][ 1: ].split('&'))
        #self.lol_name       = urllib.unquote_plus( params[ "lol_name" ] )
        #self.lol_url        = urllib.unquote_plus( params[ "lol_url" ] )
        self.url = url
        xbmc.log('URL: %s' % self.url)
        self.get_current_lols()

        
    def get_current_lols( self ) :
        #
        # Get HTML page...
        #
        feed = feedparser.parse(self.url)
        for entry in feed.entries:
            title       = entry.title
            if "VIDEO:" in title: continue
            
            content = entry.content[0].value
            images = BeautifulStoneSoup(content).findAll('img')
            full_image_url = ''

            if ".jpg" not in images[0]['src']: continue
            full_image_url = images[0]['src']
            if full_image_url == '': continue
         
            thumbnail_url = full_image_url

            # Add directory entry...
            listitem = xbmcgui.ListItem( title, iconImage="DefaultPicture.png", thumbnailImage = thumbnail_url )
            xbmcplugin.addDirectoryItem( handle=int(sys.argv[ 1 ]), url = full_image_url, listitem=listitem, isFolder=False)

        #
        # Next page entry...
        #
        #next_button_text = __language__(30403) % (self.entries_per_page, self.lol_name)
        #listitem      = xbmcgui.ListItem (next_button_text, iconImage = "DefaultFolder.png", thumbnailImage = os.path.join(self.IMAGES_PATH, 'next-page.png'))
        #next_page_url = "%s?action=list&lol_name=%s&lol_url=%s" % ( sys.argv[0], urllib.quote_plus( self.lol_name), urllib.quote_plus( self.lol_url ) )
        #next_page_url = self.lol_url
        #xbmcplugin.addDirectoryItem( handle = int(sys.argv[1]), url = next_page_url, listitem = listitem, isFolder = True)
        if __settings__.getSetting("home_page") == "1":
            # The default start page is this page, so add a Home Page item
            next_button_text = __language__(30211)
            listitem = xbmcgui.ListItem(next_button_text, iconImage = "DefaultFolder.png", thumbnailImage="DefaultFolder.png")
            url = "%s?action=forcemain" % sys.argv[0]
            xbmcplugin.addDirectoryItem(handle=int(sys.argv[1]), url=url, listitem=listitem, isFolder=True)
        #
        # Disable sorting...
        #
        xbmcplugin.addSortMethod( handle=int( sys.argv[ 1 ] ), sortMethod=xbmcplugin.SORT_METHOD_NONE )

        #
        # Label (top-right)...
        #
        #trlabel = "%s" % self.lol_name
        #xbmcplugin.setPluginCategory( handle=int( sys.argv[ 1 ] ), category=( trlabel ) )

        #
        # End of directory...
        #
        xbmcplugin.endOfDirectory( handle=int( sys.argv[ 1 ] ), succeeded=True )



resources/lib/cheezburger_type_selection.py
# To change this template, choose Tools | Templates
# and open the template in the editor.

__author__="brian"
__date__ ="$Aug 20, 2010 10:36:42 PM$"

import sys
import xbmcgui
import xbmcplugin
import urllib

import xbmcaddon

__settings__ = xbmcaddon.Addon(id='plugin.pictures.icanhascheezburger.com')
__language__ = __settings__.getLocalizedString

#
# Main class
#
class Main:
    #
    # Init
    #
    def __init__(self):
        #
        # Get Lols
        #
        self.get_cheez_type( )

    def get_cheez_type(self):
        cheez_type = {}
        for t in range(30405, 30407):
            cheez_type[t] = __language__(t)
            plugin_url = "%s?action=%s" % (
             sys.argv[0],
             t )
            listitem = xbmcgui.ListItem (__language__(t),
             iconImage = "DefaultPicture.png",
             thumbnailImage = "DefaultPicture.png")
            listitem.setInfo( "pictures", { "title" : "lol_name" } )
            ok = xbmcplugin.addDirectoryItem( handle = int(sys.argv[1]),
             url = plugin_url,
             listitem = listitem,
             isFolder = True)

        #
        # End of list...
        #

        # Sort by label...
        xbmcplugin.addSortMethod( handle=int( sys.argv[ 1 ] ), sortMethod=xbmcplugin.SORT_METHOD_LABEL )

        # End of list...
        xbmcplugin.endOfDirectory( handle=int( sys.argv[ 1 ] ), succeeded=True )


resources/lib/current_cheez.py
__author__="brian"
__date__ ="$Aug 20, 2010 11:11:39 PM$"

# Show current LOLs. For now, only Cats are supported. Others will follow.

import os
import re
import sys
import xbmc
import xbmcgui
import xbmcplugin
import feedparser
from BeautifulSoup    import BeautifulStoneSoup
import xbmcaddon

__settings__ = xbmcaddon.Addon(id='plugin.pictures.icanhascheezburger.com')
__language__ = __settings__.getLocalizedString

class Main:
    #
    # Init
    #
    def __init__( self ) :
        # Constants
        self.DEBUG            = False
        self.IMAGES_PATH      = xbmc.translatePath( os.path.join( os.getcwd(), 'resources', 'images' ) )

        self.entries_per_page = 5

        if __settings__.getSetting ("entries_per_page") == "0" :
            self.entries_per_page = 5
        elif __settings__.getSetting ("entries_per_page") == "1" :
            self.entries_per_page = 10
        elif __settings__.getSetting ("entries_per_page") == "2" :
            self.entries_per_page = 20

        self.get_current_lols()

    def get_current_lols( self ) :
        #
        # Get HTML page...
        #
        self.lol_name = "LOL"
        self.current_page = 1
        self.lol_url = "URL"
        url_base = "http://feeds.feedburner.com/ICanHasCheezburger"
        feed = feedparser.parse(url_base)
        for entry in feed.entries:
            title       = entry.title
            if "VIDEO:" in title: continue
            
            content = entry.content[0].value
            images = BeautifulStoneSoup(content).findAll('img')
            full_image_url = ''

            if ".jpg" not in images[0]['src']: continue
            full_image_url = images[0]['src']
            if full_image_url == '': continue
         
            thumbnail_url = full_image_url

            # Add directory entry...
            listitem = xbmcgui.ListItem( title, iconImage="DefaultPicture.png", thumbnailImage = thumbnail_url )
            xbmcplugin.addDirectoryItem( handle=int(sys.argv[ 1 ]), url = full_image_url, listitem=listitem, isFolder=False)

        #
        # Next page entry...
        #
        #next_button_text = __language__(30403) % (self.entries_per_page, self.lol_name)
        #listitem      = xbmcgui.ListItem (next_button_text, iconImage = "DefaultFolder.png", thumbnailImage = os.path.join(self.IMAGES_PATH, 'next-page.png'))
        #next_page_url = "%s?action=list&lol_name=%s&lol_url=%s" % ( sys.argv[0], urllib.quote_plus( self.lol_name), urllib.quote_plus( self.lol_url ) )
        #next_page_url = self.lol_url
        #xbmcplugin.addDirectoryItem( handle = int(sys.argv[1]), url = next_page_url, listitem = listitem, isFolder = True)
        if __settings__.getSetting("home_page") == "1":
            # The default start page is this page, so add a Home Page item
            next_button_text = __language__(30211)
            listitem = xbmcgui.ListItem(next_button_text, iconImage = "DefaultFolder.png", thumbnailImage="DefaultFolder.png")
            url = "%s?action=forcemain" % sys.argv[0]
            xbmcplugin.addDirectoryItem(handle=int(sys.argv[1]), url=url, listitem=listitem, isFolder=True)
        #
        # Disable sorting...
        #
        xbmcplugin.addSortMethod( handle=int( sys.argv[ 1 ] ), sortMethod=xbmcplugin.SORT_METHOD_NONE )

        #
        # Label (top-right)...
        #
        trlabel = "%s (%s)" % ( self.lol_name, (__language__(30402) % self.current_page))
        xbmcplugin.setPluginCategory( handle=int( sys.argv[ 1 ] ), category=( trlabel ) )

        #
        # End of directory...
        #
        xbmcplugin.endOfDirectory( handle=int( sys.argv[ 1 ] ), succeeded=True )

